### PR TITLE
Added scenario tests for MessageContractAttribute and  ServiceKnownTypeAttribute.

### LINKED ISSUE #557: Add Scenario Tests Covering Using XmlSerializerFormat with MessageContract

### LINKED ISSUE #569: Add Scenario Tests for ServiceKnownTypeAttribute
We should add tests for ServiceKnownTypeAttribute. We also need tests covering the scenario where ServiceKnownTypeAttribute is used together with XmlSerializerFormatAttribute.

### PR DESCRIPTION
**Pull request by shmao**
Added a scenario test for XmlSerializerFormat & MessageContract.
Added scenario tests for ServiceKnownTypeAttribute.

Fix #557.
Fix #569.

### CODE REVIEW

**Comment by shmao** [comment]
@dotnet-bot test outerloop please

File: src/System.Private.ServiceModel/tests/Scenarios/Contract/Service/ServiceKnownTypeTests.cs [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using Xunit;
+
+
+public static class ServiceKnownTypeTests
+{
+    [Fact]
+    [OuterLoop]
+    public static void ServiceKnownType_DataContract_AttrOnMethod_Test()
```
Comment (StephenBonikowsky) [inline src/System.Private.ServiceModel/tests/Scenarios/Contract/Service/ServiceKnownTypeTests.cs]: I know this is annoying but could you add the comments we are now using with all new tests to call out the distinct sections of the test. One of the reasons for insisting on this is that we are hoping to use our test bed as examples for the community on how to use WCF and we want them to have a consistent style and syntax. See tests in other areas for guidance such as Scenarios\Contract\Service\ServiceContractTests.cs.
// *** SETUP *** \\
// *** EXECUTE *** \\
// *** VALIDATE *** \\
// *** CLEANUP *** \\
// *** ENSURE CLEANUP *** \\

File: src/System.Private.ServiceModel/tests/Scenarios/Contract/Service/ServiceKnownTypeTests.cs [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+        {
+            var widget = new Widget0();
+            widget.Id = "1";
+            widget.Catalog = "widget";
+
+            var input = new object[] { widget };
+            var response = serviceProxy.EchoItems(input);
+
+            Assert.Equal(widget.Id, ((Widget)response[0]).Id);
+        }
```
Comment (StephenBonikowsky) [inline src/System.Private.ServiceModel/tests/Scenarios/Contract/Service/ServiceKnownTypeTests.cs]: As the last thing before the Finally add some lines to attempt closing of any communication objects...

// *** CLEANUP *** \\
            ((ICommunicationObject)serviceProxy).Close();
            factory.Close();

File: src/System.Private.ServiceModel/tests/Scenarios/Contract/Service/ServiceKnownTypeTests.cs [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        finally
+        {
```
Comment (StephenBonikowsky) [inline src/System.Private.ServiceModel/tests/Scenarios/Contract/Service/ServiceKnownTypeTests.cs]: In the Finally just use the helper method...

// *** ENSURE CLEANUP *** \\
            ScenarioTestHelpers.CloseCommunicationObjects((ICommunicationObject)serviceProxy, factory);

File: src/System.Private.ServiceModel/tests/Scenarios/Contract/Service/ServiceKnownTypeTests.cs [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+        try
+        {
+            var widget = new Widget1();
+            widget.Id = "1";
+            widget.Catalog = "widget";
+
+            var input = new object[] { widget };
+            var response = serviceProxy.EchoItems(input);
+
+            Assert.Equal(widget.Id, ((Widget)response[0]).Id);
```
Comment (StephenBonikowsky) [inline src/System.Private.ServiceModel/tests/Scenarios/Contract/Service/ServiceKnownTypeTests.cs]: Can you include a message in the case of a failure that makes some reference to 'actual' and 'expected'?

File: src/System.Private.ServiceModel/tests/Scenarios/Contract/Service/ServiceKnownTypeTests.cs [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+        try
+        {
+            var widget = new Widget2();
+            widget.Id = "1";
+            widget.Catalog = "widget";
+
+            var input = new object[] { widget };
+            var response = serviceProxy.EchoItems_Xml(input);
+
+            Assert.Equal(widget.Id, ((Widget)response[0]).Id);
```
Comment (StephenBonikowsky) [inline src/System.Private.ServiceModel/tests/Scenarios/Contract/Service/ServiceKnownTypeTests.cs]: Include error message. (Same as prior test)

File: src/System.Private.ServiceModel/tests/Scenarios/Contract/Service/ServiceKnownTypeTests.cs [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+        try
+        {
+            var widget = new Widget3();
+            widget.Id = "1";
+            widget.Catalog = "widget";
+
+            var input = new object[] { widget };
+            var response = serviceProxy.EchoItems_Xml(input);
+
+            Assert.Equal(widget.Id, ((Widget)response[0]).Id);
```
Comment (StephenBonikowsky) [inline src/System.Private.ServiceModel/tests/Scenarios/Contract/Service/ServiceKnownTypeTests.cs]: Include error message. (Same as prior test)

File: src/System.Private.ServiceModel/tests/Scenarios/Contract/XmlSerializer/XmlSerializerFormatTest.cs [hunk 7]
```diff
@@ -107,4 +107,23 @@ public static void XmlSerializerFormat_RoundTrips_CompositeType()
...
+        ILoginService serviceProxy = factory.CreateChannel();
+
+
+        var request = new LoginRequest();
+        request.clientId = "1";
+        request.user = "2";
+        request.pwd = "3";
+        var response = serviceProxy.Login(request);
+        Assert.True(response != null);
+        Assert.Equal("123", response.@return);
```
Comment (StephenBonikowsky) [inline src/System.Private.ServiceModel/tests/Scenarios/Contract/XmlSerializer/XmlSerializerFormatTest.cs]: Include error message. (Same as prior tests)

**Comment by StephenBonikowsky** [comment]
LGTM after comments are addressed.

File: src/System.Private.ServiceModel/tests/Scenarios/Contract/Service/ServiceKnownTypeTests.cs [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+            {
+                factory.Abort();
+            }
+        }
+    }
+
+    [Fact]
+    [OuterLoop]
+    public static void ServiceKnownType_XmlSerializerFormat_AttrOnMethod_Test()
+    {
```
Comment (roncain) [inline src/System.Private.ServiceModel/tests/Scenarios/Contract/Service/ServiceKnownTypeTests.cs]: Won't you need to add [ActiveIssue] to the Xml tests before checkin?  I believe these tests will fail until the toolchain is fixed.

**Comment by roncain** [comment]
LGTM, subject to the cleanups @StephenBonikowsky suggested

File: src/System.Private.ServiceModel/tests/Scenarios/Contract/Service/ServiceKnownTypeTests.cs (hunk 1 above)
Comment (shmao) [inline src/System.Private.ServiceModel/tests/Scenarios/Contract/Service/ServiceKnownTypeTests.cs]: I'll add these in the tests. Thanks.

File: src/System.Private.ServiceModel/tests/Scenarios/Contract/Service/ServiceKnownTypeTests.cs (hunk 4 above)
Comment (shmao) [inline src/System.Private.ServiceModel/tests/Scenarios/Contract/Service/ServiceKnownTypeTests.cs]: Assert.Equal does not have an overload accepting a error message as parameter. And it should by default provide an error message with references to 'actual' and 'expected'.

File: src/System.Private.ServiceModel/tests/Scenarios/Contract/Service/ServiceKnownTypeTests.cs (hunk 8 above)
Comment (shmao) [inline src/System.Private.ServiceModel/tests/Scenarios/Contract/Service/ServiceKnownTypeTests.cs]: Yes, correct.

**Comment by shmao** [comment]
@dotnet-bot test outerloop please

**Comment by shmao** [comment]
@dotnet-bot test outerloop please

**Comment by shmao** [comment]
@dotnet-bot test outerloop please

**Pull request merged**